### ISSUE TITLE
wrong binds types passed from REST/Server.py to cx_oracle due to py23 changes

### ISSUE THREAD

**Issue by belforte** [dmwm/WMCore#10335]
**Impact of the bug**
CRAB (which AFAIK is the only user of WMCore/REST as a REST server on top of Oracle DB)

**Describe the bug**
It is tricky to describe since it is not fully understood yet. 
This is mostly a **heads up** in case @mapellidario has already an idea.

Please take this as a general report aimed at getting your (e.g. @amaltaro @mapellidario ) help in 
defining a good strategy for CRAB migration. I do not like to stick with an old 1.3.crab branch longer,
but the fact that we are the only user of this class, which has no testing in WMCore fwk, makes things tricky
https://developer.oracle.com/dsl/tuininga-cx-oracle.html

For a while @ddaina and myself have been chasing the fact that just some queries at times were taking O(minutes) in preprod,
while all queries were fast on our dev DB instance after we moved to use WMCore with py23 compat (1.4.6.pre2).
We ruled out a change in dependencies (cx_oracle e.g.) and could confirm by simply picking python from
a different GH tag inside a built image.
But faced with :
1. old image is fast on prod, preprod, dev
2. new image is slow on prod, preprod but fast on dev
we could not isolate problem to our code, rather than DB.

Eventually Katarzyna Maria Dziedziniewicz-Wojcik (aka Kate) notices that Oracle optimizer was using full table scale,
in which case the much smaller table size in "dev" DB may explain the difference.

Log inspection showed that queries like
`UPDATE blah WHERE id = :var` with `kwbinds={'var',:'somestring'}`
where instead being executed as
`UPDATE blah WHERE id = :var` with `kwbinds={'var',:u'somestring'}`
the unicode in the `WHERE` is making Oracle fall to full table scan (do you remember the DBS vs. new SQLAlchemy issue 2 in spring 2019 ?).

I stepped through things with pdb, and the time is lost exactly in
https://github.com/dmwm/WMCore/blob/935ef0de5404c0fa10a41b62bede25233baa6c3f/src/python/WMCore/REST/Server.py#L2098
or the equivalent
https://github.com/dmwm/WMCore/blob/935ef0de5404c0fa10a41b62bede25233baa6c3f/src/python/WMCore/REST/Server.py#L2098
But I can't follow code inside the `c.execute` 

Long story short, and much confusion having  flown and still reigning in my head, Kate has now
indicated that we can force the type of the bind variable with code like

```
binds = ()
k='id'
kwbinds = {k:'2b69a73edc024c216dc285fdee8b9044a9c3de784355fa341104eec2'}
sizes = {k: cx_Oracle.STRING}
cur.setinputsizes(**sizes)
#cur.execute(None, id = s)
cur.execute(None, *binds, **kwbinds)
```
I still have to test this, but a more general solution may be needed, since at times
the varaibles used in the `WHERE` are NUMBERS not all VARCHAR.
Kate suggested to look here : https://developer.oracle.com/dsl/tuininga-cx-oracle.html , I still have to digest that.

It is also very much possible that all of this disappears by simply running with python3 !!

**How to reproduce it**
tricky

**Expected behavior**
Oracle should never use full table scan and queries should be completed in msec

**Additional context and error message**

**Comment by belforte** [comment]
Kate said:
Let me put it this way - we've spent 40mins yesterday trying to install python 2 with proper cx_oracle and instant client
and we have only managed because we have valgrind  and some oooold ubuntu images and a collection of instact client libraries downloaded to our very fast home pc, otherwise it's not possible in a reasonable amount of time
move to 3 when you can  - fixing unsupported 2 is really a pain now

**Comment by amaltaro** [comment]
@belforte Stefano, I haven't yet looked very carefully at what you're reporting here. However, can you please clarify what the impact and how important it is to get this fixed before the migration to python3?

I second what Kate said (and I'm pretty sure Dario too), but depending on your answer to these questions above, we might have to work on some temporary workaround... speaking of which, a long time ago we implemented this fix for DBS to deal with issues that popped up with a cx-oracle upgrade:
https://github.com/dmwm/WMCore/pull/9208

it never got merged because there is, of course, a performance penalty to cast every SQL query, and every response; but it might be an option here.

**Comment by belforte** [comment]
thanks @amaltaro . Let's take breath. I think it is important now for me and Daina to understand where we are and define some form points. And I think all changes to WMCore, even tentative, should be tracke here, not in CRABServer repo.
Once we are on solid ground, we should think.
Impact: I have been foolish enough to think that everything was working and now it is easier for us to stay in the WMCore 1.4.6 branch. But with a couple days of work we can go back to 1.3.crab.6 and use that for the project of the months (replace CRABCache with S3), and look at py3.
I think at this point we only need guidance from you on which work to do in which environment, and when.

**Comment by belforte** [comment]
P.S. I had tried s cruder version of what you have in #9208, but that only got our queries from 80 to 3 seconds. Then Kate suggested to enforce cx_Oracle.STRING type for the bind variables. There's also some "persistency" on Oracle side (once the optimizer decided to change execution plan, may stick with it for a while no matter what) which makes things not precisely reproducible... let's not go there !

**Comment by belforte** [comment]
as indicated in the the referred-to CRABServer issue, I have implemented Kate's recipe in my branch https://github.com/belforte/WMCore/tree/OraFixes and so far is working.

**Comment by belforte** [comment]
I have created a branch off 1.4.6 in my fork and put there a possible fix following suggestion by Kate
https://github.com/belforte/WMCore/tree/OraFixes
the relevant change is https://github.com/belforte/WMCore/commit/85ec106f5a2cb4e6076dbe46281511d3459fef3d
A crabserver image created with that is working great in cmsweb-testbed and cmsweb-test2
But I do not trust my code to be good enough to cover all possible use case, even if it works for CRAB and CRAB is the only user of the modified class
https://github.com/dmwm/WMCore/blob/595a9934aef084b3aaf5ca93c55a145c405edb43/src/python/WMCore/REST/Server.py#L1666

So I will **not** propose to merge that into WMCore master, but rather wait to see what things are once we run this in python3.
Expectation is that python3 version of cx_oracle will handle unicode graciously and automatically cast strings into `cx_Oracle.STRING` as needed.
 
We should define a good strategy for keep working and updating CRAB in the immediate future while working also on py3 migration. @ddaina FYI

**Comment by belforte** [comment]
@amaltaro here's proposal on how to progress:

1. create another crab branch in WMCore, forking from latest `1.4.6`  e.g. `1.4.6.crab`
2. port to that the PR about REST/Main debug mode : https://github.com/dmwm/WMCore/pull/10339
3. then I will make a PR to put the changes to REST/Server.py discussed here ONLY to 1.4.6.crab branch
4. we will keep using WMCore from there for the next few months while you go on with python3 work
5. once there is a python3 working environment in COMP we will start testing CRAB with python3
6. hopefully all changes in WMCore from now to then will be in parts of WMCore that we do not use
7. so when using python3 we will build with current WMCore at that time and hopefully there will be no need of the REST/Server.py hack discussed in this, since py3 version of cx_oracle will nicely take care of unicode, newbytes etc.

Of course we'll refine last steps as we go a will figure out things when we hit them. But if overall sounds good to you as well, please go ahead with 1. and 2. at your convenience.

Thanks

**Comment by amaltaro** [comment]
@belforte Stefano, thanks for this proposing this. It sounds like a good plan.

1. branch `1.4.6_crab` has been created
2. backport of #10339 is provided here: https://github.com/dmwm/WMCore/pull/10350

Please let me know what the next steps are.

**Comment by belforte** [comment]
thanks Alan. I will verify 1.4.6_crab and make a PR for REST/Server.py (point 3. in the plan). Then @ddaina will build a crabserver image with this, test and put in production. We will use this as base for py3 evolution.

**Comment by belforte** [comment]
fixed in branch 1.4.6_crab via https://github.com/dmwm/WMCore/pull/10351
